src/lambda/wizards/samDeployWizard.ts

```diff
@@ -299,35 +299,47 @@ export class DefaultSamDeployWizardContext implements SamDeployWizardContext {
     /**
      * Retrieves an S3 Bucket to deploy to from the user.
      *
+     * @param selectedRegion Selected region for S3 client usage
      * @param initialValue Optional, Initial value to prompt with
+     * @param messages Passthrough strings for testing
      *
      * @returns S3 Bucket name. Undefined represents cancel.
      */
-    public async promptUserForS3Bucket(selectedRegion: string, initialValue?: string): Promise<string | undefined> {
-        const inputBox = input.createInputBox({
+    public async promptUserForS3Bucket(
+        selectedRegion: string,
+        initialValue: string | undefined = undefined,
+        messages: {
+            noBuckets: string
+            bucketError: string
+        } = {
+            noBuckets: localize('AWS.samcli.deploy.s3bucket.picker.noBuckets', 'No buckets found.'),
+            bucketError: localize('AWS.samcli.deploy.s3bucket.picker.error', 'There was an error loading S3 buckets.'),
+        }
+    ): Promise<string | undefined> {
+        const loadingBuckets: string = localize('AWS.samcli.deploy.s3bucket.picker.loading', 'Loading S3 buckets...')
+
+        const quickPick = picker.createQuickPick<vscode.QuickPickItem>({
             buttons: [this.helpButton, vscode.QuickInputButtons.Back],
             options: {
-                title: localize(
-                    'AWS.samcli.deploy.s3Bucket.prompt',
-                    'Enter the AWS S3 bucket to which your code should be deployed'
-                ),
+                title: localize('AWS.samcli.deploy.s3Bucket.prompt', 'Select an AWS S3 Bucket to deploy code to'),
+                value: initialValue,
+                matchOnDetail: true,
                 ignoreFocusOut: true,
-                prompt: localize(
-                    'AWS.samcli.deploy.s3Bucket.region',
-                    'S3 bucket must be in selected region: {0}',
-                    selectedRegion
-                ),
             },
+            items: [
+                {
+                    label: loadingBuckets,
+                },
+            ],
         })
 
-        // Pre-populate the value if it was already set
-        if (initialValue) {
-            inputBox.value = initialValue
-        }
+        quickPick.busy = true
+        quickPick.enabled = false
 
-        return await input.promptUser({
-            inputBox: inputBox,
-            onValidateInput: validateBucketName,
+        populateS3QuickPick(quickPick, selectedRegion, messages)
```

Comment:
Add a comment here stating that we aren't waiting for the s3 bucket load to complete, so future maintainers know that something is happening beyond this call